packages/shared/src/worker/worker.ts

```diff
@@ -0,0 +1,168 @@
+import type { ProgressInfo } from '@huggingface/transformers'
+
+import type { LoadOptionProgressCallback } from '../types'
+
+export type LoadMessageEvents<D = undefined, T extends string = string> =
+  | WorkerMessageEvent<D, T>
+  | WorkerMessageEvent<{ error?: unknown, message?: string }, 'error'>
+  | WorkerMessageEvent<{ message: string }, 'info'>
+  | WorkerMessageEvent<{ message?: string, status: 'loading' | 'ready' }, 'status'>
+  | WorkerMessageEvent<{ progress: ProgressInfo }, 'progress'>
+
+export type ProcessMessageEvents<D = unknown, T = unknown> =
+  | WorkerMessageEvent<D, T>
+
+export interface WorkerMessageEvent<D, T> {
+  data: D
+  type: T
+}
+
+export type WorkerMessageEvents<D = undefined, T extends string = string> =
+  | LoadMessageEvents<D, T>
+  | ProcessMessageEvents<D, T>
+
+export const createTransformersWorker = <
+  T extends { workerURL: string | undefined | URL },
+  T2 extends { onProgress?: LoadOptionProgressCallback },
+>(createOptions: T) => {
+  let worker: undefined | Worker
+  let isReady = false
+  let isLoading = false
+
+  const load = async <D = unknown, E extends WorkerMessageEvent<D, string> = WorkerMessageEvent<D, string>>(payload?: E, options?: T2) => {
+    if (!payload)
+      throw new Error('Payload is required')
+
+    // eslint-disable-next-line sonarjs/cognitive-complexity
+    return new Promise<void>((resolve, reject) => {
+      if (!createOptions.workerURL)
+        throw new Error('Worker URL is required')
+
+      if (isReady) {
+        resolve()
+        return
+      }
+
+      try {
+        let onProgress: LoadOptionProgressCallback | undefined
+        if (options != null && 'onProgress' in options && options.onProgress != null) {
```

Comment:
Curious: Could this be simplified? (Same for similar cases)

```suggestion
        if (options?.onProgress) {
```

Reply:
Oh yes surely can. I think this was left over because of the missing types in early development time.